cfme/common/provider.py

```diff
@@ -404,12 +426,20 @@ def _do_stats_match(self, client, stats_to_match=None, refresh_timer=None, ui=Fa
 
     @property
     def exists(self):
-        ems = cfmedb()['ext_management_systems']
-        provs = (prov[0] for prov in cfmedb().session.query(ems.name))
-        if self.name in provs:
+        """ Checks if exists among managed providers
+
+        Note:
+            If provider exists but with a different name, name is updated
+        """
+        for prov in store.current_appliance.managed_providers:
```

Comment:
Not sure I follow. Do you propose that `exists` should have multiple implementations defined here eventually instead of having multiple implementations of `managed_providers`? That should be scalable.

Use case 1: I want to know if a provider exists and I don't care where the info comes from => I call `exists` with the default context and that's it.

Use case 2: I want to know if a provider exists and I want to check that info through REST => I call `exists` with REST context. It can either pass through to managed_providers with REST context or we can override it by defining a specific way of getting that information directly in `exists`, through REST. If none of those are available, it should raise an exception.

Still not sure I understand what you meant though, maybe that's not what you meant...